src/pages/achievement/dashboard/subcomponents/utils/AchievementInferencer.ts

```diff
@@ -0,0 +1,403 @@
+import assert from 'assert';
+
+import {
+  AchievementGoal,
+  AchievementItem,
+  AchievementStatus,
+  FilterStatus
+} from '../../../../../features/achievement/AchievementTypes';
+
+/**
+ * An InferencerNode item encapsulates all important information of an achievement item
+ *
+ * @param {AchievementItem} achievement the achievement item
+ * @param {number} dataIdx the key to retrive the achivement item in achievements[], i.e. achievements[dataIdx] = achievement
+ * @param {number} displayExp total achievable EXP of the achievement
+ * @param {number} progressFrac progress percentage in fraction. It is always between 0 to 1, both inclusive.
+ * @param {AchievementStatus} status the achievement status
+ * @param {Date | undefined} displayDeadline deadline displayed on the achievement card
+ * @param {Set<number>} children a set of immediate prerequisites id
+ * @param {Set<number>} descendant a set of all descendant prerequisites id (including immediate prerequisites)
+ */
+class InferencerNode {
+  public achievement: AchievementItem;
+  public dataIdx: number;
+  public displayExp: number;
+  public progressFrac: number;
+  public status: AchievementStatus;
+  public displayDeadline?: Date;
+  public children: Set<number>;
+  public descendant: Set<number>;
+
+  constructor(achievement: AchievementItem, dataIdx: number) {
+    const { deadline, prerequisiteIds, goals } = achievement;
+
+    this.achievement = achievement;
+    this.dataIdx = dataIdx;
+    this.displayExp = this.generateDisplayExp(goals);
+    this.progressFrac = this.generateProgressFrac(goals);
+    this.status = AchievementStatus.ACTIVE; // to be updated after the nodeList is constructed
+    this.displayDeadline = deadline;
+    this.children = new Set(prerequisiteIds);
+    this.descendant = new Set(prerequisiteIds);
+  }
+
+  private generateDisplayExp(goals: AchievementGoal[]) {
+    return goals.reduce((exp, goal) => exp + goal.goalTarget, 0);
+  }
+
+  private generateProgressFrac(goals: AchievementGoal[]) {
+    const progress = goals.reduce((progress, goal) => progress + goal.goalProgress, 0);
+
+    return Math.min(progress / this.displayExp, 1);
+  }
+}
+
+class AchievementInferencer {
```

Comment:
Please put comments on the rationale behind this class.

Reply:
👍